backend/chainlit/data/chainlit_data_layer.py

```diff
@@ -0,0 +1,569 @@
+            type="file",
+            url=str(row.get("url")),
+            name=str(row.get("name")),
+            mime=str(row.get("mime")),
+            objectKey=str(row.get("objectKey")),
+            forId=str(row.get("stepId")),
+            chainlitKey=row.get("chainlitKey"),
+            display=row["display"],
+            size=row["size"],
+            language=row["language"],
+            page=row["page"],
+            autoPlay=row.get("autoPlay"),
+            playerConfig=row.get("playerConfig"),
+        )
+
+    @queue_until_user_message()
+    async def delete_element(self, element_id: str, thread_id: Optional[str] = None):
+        """Does not delete the actual file from storage."""
+
+        query = """
+        DELETE FROM "Attachment" 
+        WHERE id = :element_id
+        """
+        if thread_id:
+            query += ' AND "threadId" = :thread_id'
+
+        await self.execute_query(
+            query, {"element_id": element_id, "thread_id": thread_id}
+        )
+
+    @queue_until_user_message()
+    async def create_step(self, step_dict: StepDict):
+        if step_dict.get("threadId"):
+            thread_query = """
+            SELECT id FROM "Thread" WHERE id = :thread_id
+            """
+            thread_result = await self.execute_query(
+                thread_query, {"thread_id": step_dict["threadId"]}
+            )
+            if not thread_result.first():
+                await self.update_thread(thread_id=step_dict["threadId"])
+
+        now = await self.get_current_timestamp()
+
+        if step_dict.get("parentId"):
+            parent_query = """
+            SELECT id FROM "Step" WHERE id = :parent_id
+            """
+            parent_result = await self.execute_query(
+                parent_query, {"parent_id": step_dict["parentId"]}
+            )
+            if not parent_result.first():
+                await self.create_step(
+                    {
+                        "id": step_dict["parentId"],
+                        "metadata": json.dumps(step_dict.get("metadata", {})),
+                        "type": "run",
+                        "start_time": now,
+                        "end_time": now,
+                        "variables": json.dumps(step_dict.get("variables", {})),
+                        "settings": json.dumps(step_dict.get("settings", {})),
+                        "tools": json.dumps(step_dict.get("tools", {})),
+                    }
+                )
+        query = """
+        INSERT INTO "Step" (
+            id, "threadId", "parentId", input, metadata, name, output,
+            type, "startTime", "endTime", variables, settings, tools,
+            "rootRunId"
+        ) VALUES (
+            :id, :thread_id, :parent_id, :input, :metadata, :name, :output, 
+            :type, :start_time, :end_time, :variables, :settings, :tools,
+            :root_run_id
+        )
+        ON CONFLICT (id) DO UPDATE SET
```

Comment:
I had to go with `on conflict` because a step might already exist in DB, just not yet populated.